org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/commands/ClasspathCommand.java

```diff
@@ -0,0 +1,204 @@
+				ArrayList<ClasspathItem> children = new ArrayList<>();
+
+				for (IClasspathEntry entry : container.getClasspathEntries()) {
+					IPackageFragmentRoot packageFragmentRoot = javaProject.findPackageFragmentRoot(entry.getPath());
+					children.add(new ClasspathItem(packageFragmentRoot.getElementName(), entry.getPath(), ClasspathItem.JAR));
+				}
+				return Either.forLeft(children.toArray(new ClasspathItem[children.size()]));
+			} catch (CoreException e) {
+				JavaLanguageServerPlugin.logException("Problem load project JAR entries ", e);
+			}
+		}
+
+		return Either.forLeft(new ClasspathItem[0]);
+	}
+
+	private static Either<ClasspathItem[], String> getPackages(List<String> query) {
+		IJavaProject javaProject = getJavaProject(query.get(0));
+
+		if (javaProject != null) {
+			try {
+				IPackageFragmentRoot packageRoot = javaProject.findPackageFragmentRoot(ResourceUtils.filePathFromURI(query.get(1)));
+				if (packageRoot == null) {
+					throw new CoreException(new Status(IStatus.ERROR, JavaLanguageServerPlugin.PLUGIN_ID, String.format("No package root found for %s", query.get(1))));
+				}
+				IJavaElement[] elements = packageRoot.getChildren();
+				return Either.forLeft(Arrays.stream(elements).filter(element -> {
+					if (element instanceof PackageFragment) {
+						try {
+							return ((PackageFragment) element).hasChildren();
+						} catch (JavaModelException e) {
+							return false;
+						}
+					}
+					return false;
+				}).map(element -> new ClasspathItem(element.getElementName(), element.getPath(), ClasspathItem.PACKAGE)).toArray(ClasspathItem[]::new));
+			} catch (CoreException e) {
+				JavaLanguageServerPlugin.logException("Problem load project package ", e);
+			}
+		}
+		return Either.forLeft(new ClasspathItem[0]);
+	}
+
+	private static Either<ClasspathItem[], String> getClassfiles(List<String> query) {
+		IJavaProject javaProject = getJavaProject(query.get(0));
+		if (javaProject != null) {
+			try {
+				IPackageFragmentRoot packageRoot = javaProject.findPackageFragmentRoot(ResourceUtils.filePathFromURI(query.get(1)));
+				if (packageRoot == null) {
+					throw new CoreException(new Status(IStatus.ERROR, JavaLanguageServerPlugin.PLUGIN_ID, String.format("No package root found for %s", query.get(1))));
+				}
+				IPackageFragment packageFragment = packageRoot.getPackageFragment(query.get(2));
+				if (packageFragment != null) {
+					IClassFile[] classFiles = packageFragment.getAllClassFiles();
+					return Either.forLeft(Arrays.stream(classFiles).map(classFile -> new ClasspathItem(classFile.getElementName(), classFile.getPath(), ClasspathItem.CLASSFILE)).toArray(ClasspathItem[]::new));
+				}
+			} catch (CoreException e) {
+				JavaLanguageServerPlugin.logException("Problem load project classfile list ", e);
+			}
+		}
+		return Either.forLeft(new ClasspathItem[0]);
+	}
+
+	private static Either<ClasspathItem[], String> getSource(List<String> query) {

```

Comment:
there's no need to reinvent a source loading mechanism. Clients should load sources from urls

Reply:
This sounds great and it will simplify both client and server logic a lot. I will try this one.

Reply:
Okay, the source file is packed insider a source jar. In this case, how to load sources from this jar?

Reply:
Use [JDTUtils.toURI(ITypeRoot)](https://github.com/eclipse/eclipse.jdt.ls/blob/aa79b5733115374ac97a32aabcaab0c947a6fac8/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java#L431) to return the source uri. If the file is in the workspace, it'll return a file URI, if it's a class, it'll return a jdt:// URI. The client will then have to request the server to open that URI, which will return the source code. Exactly the same mechanism used for navigating code.

Reply:
Fixed.